File: lc_soln/0287_find_the_duplicate_number/solution.cs
public class Solution
{
    public int FindDuplicate(int[] nums)
    {
        int fast = nums[0];
        int slow = nums[0];
        int result = -1;
        while (true)
        {
            fast = nums[nums[fast]];
            slow = nums[slow];
            if (fast == slow)
            {
                int slow2 = nums[0];
                while (slow != slow2)
                {
                    slow = nums[slow];
                    slow2 = nums[slow2];
                }
                result = slow;
                break;
            }
        }
        return result;
    }
}

File: lc_soln/0287_find_the_duplicate_number/solution.rs
struct Solution;

impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let mut fast = nums[0];
        let mut slow = nums[0];
        let mut result = -1;
        loop {
            fast = nums[nums[fast as usize] as usize];
            slow = nums[slow as usize];
            if fast == slow {
                let mut slow2 = nums[0];
                while slow != slow2 {
                    slow = nums[slow as usize];
                    slow2 = nums[slow2 as usize];
                }
                result = slow2;
                break;
            }
        }
        result
    }
}

fn main() {}
